// @cloudscape-design/collection-hooks/cjs/operations/index.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.itemsAreEqual = exports.processSelectedItems = void 0;
exports.processItems = processItems;
const filter_js_1 = require("./filter.js");
const property_filter_js_1 = require("./property-filter.js");
const sort_js_1 = require("./sort.js");
const pagination_js_1 = require("./pagination.js");
const compose_filters_js_1 = require("./compose-filters.js");
const trackby_utils_js_1 = require("./trackby-utils.js");
const items_tree_js_1 = require("./items-tree.js");
function processItems(allItems, { filteringText, sortingState, currentPageIndex, propertyFilteringQuery }, { filtering, sorting, pagination, propertyFiltering, expandableRows }) {
    const filterPredicate = (0, compose_filters_js_1.composeFilters)((0, property_filter_js_1.createPropertyFilterPredicate)(propertyFiltering, propertyFilteringQuery), (0, filter_js_1.createFilterPredicate)(filtering, filteringText));
    const sortingComparator = (0, sort_js_1.createComparator)(sorting, sortingState);
    const { items, size, getChildren } = expandableRows
        ? (0, items_tree_js_1.computeTreeItems)(allItems, expandableRows, filterPredicate, sortingComparator)
        : (0, items_tree_js_1.computeFlatItems)(allItems, filterPredicate, sortingComparator);
    const filteredItemsCount = filterPredicate ? size : undefined;
    const pageProps = (0, pagination_js_1.createPageProps)(pagination, currentPageIndex, items);
    if (pageProps) {
        return {
            items: items.slice((pageProps.pageIndex - 1) * pageProps.pageSize, pageProps.pageIndex * pageProps.pageSize),
            allPageItems: items,
            filteredItemsCount,
            pagesCount: pageProps === null || pageProps === void 0 ? void 0 : pageProps.pagesCount,
            actualPageIndex: pageProps === null || pageProps === void 0 ? void 0 : pageProps.pageIndex,
            getChildren,
        };
    }
    return {
        items: items,
        allPageItems: items,
        filteredItemsCount,
        pagesCount: undefined,
        actualPageIndex: undefined,
        getChildren,
    };
}
const processSelectedItems = (items, selectedItems, trackBy) => {
    const selectedSet = new Set();
    selectedItems.forEach(item => selectedSet.add((0, trackby_utils_js_1.getTrackableValue)(trackBy, item)));
    return items.filter(item => selectedSet.has((0, trackby_utils_js_1.getTrackableValue)(trackBy, item)));
};
exports.processSelectedItems = processSelectedItems;
const itemsAreEqual = (items1, items2, trackBy) => {
    if (items1.length !== items2.length) {
        return false;
    }
    const set1 = new Set();
    items1.forEach(item => set1.add((0, trackby_utils_js_1.getTrackableValue)(trackBy, item)));
    return items2.every(item => set1.has((0, trackby_utils_js_1.getTrackableValue)(trackBy, item)));
};
exports.itemsAreEqual = itemsAreEqual;
